File: algorithms/micali-vazirani/python/micali_vazirani.py
"""
Micali-Vazirani Algorithm (Hybrid) - O(E√V) Maximum Matching

Hybrid approach:
- MV-style MIN phase (level building with even/odd tracking)
- Gabow-style MAX phase (simple path finding)

Python implementation — fully deterministic, no hash containers.
"""

import sys
import time
from bisect import bisect_left

NIL = -1
UNSET = float('inf')


class Node:
    __slots__ = ('preds', 'match', 'min_level', 'even_level', 'odd_level')

    def __init__(self):
        self.preds = []
        self.match = NIL
        self.min_level = UNSET
        self.even_level = UNSET
        self.odd_level = UNSET

    def set_min_level(self, level):
        self.min_level = level
        if level % 2:
            self.odd_level = level
        else:
            self.even_level = level

    def reset(self):
        self.preds.clear()
        self.min_level = UNSET
        self.even_level = UNSET
        self.odd_level = UNSET


class MicaliVazirani:
    def __init__(self, n, edges):
        self.n = n
        self.graph = [[] for _ in range(n)]
        self.nodes = [Node() for _ in range(n)]
        self.base = list(range(n))
        self.levels = []

        for u, v in edges:
            if u < n and v < n and u != v:
                self.graph[u].append(v)
                self.graph[v].append(u)
        for adj in self.graph:
            adj.sort()

    def find_base(self, v):
        if self.base[v] != v:
            self.base[v] = self.find_base(self.base[v])
        return self.base[v]

    def add_to_level(self, level, node):
        while len(self.levels) <= level:
            self.levels.append([])
        self.levels[level].append(node)

    def step_to(self, to, frm, level):
        next_level = level + 1
        tl = self.nodes[to].min_level
        if tl >= next_level:
            if tl != next_level:
                self.add_to_level(next_level, to)
                self.nodes[to].set_min_level(next_level)
            self.nodes[to].preds.append(frm)

    def phase_1(self):
        self.levels.clear()
        for i in range(self.n):
            self.base[i] = i
            self.nodes[i].reset()

        for i in range(self.n):
            if self.nodes[i].match == NIL:
                self.add_to_level(0, i)
                self.nodes[i].set_min_level(0)

        for i in range(self.n):
            if i >= len(self.levels) or not self.levels[i]:
                continue
            level_snap = list(self.levels[i])
            for cur in level_snap:
                if i % 2 == 0:
                    for nb in self.graph[cur]:
                        if nb != self.nodes[cur].match:
                            self.step_to(nb, cur, i)
                else:
                    if self.nodes[cur].match != NIL:
                        self.step_to(self.nodes[cur].match, cur, i)

    def phase_2(self):
        found = False

        for start in range(self.n):
            if self.nodes[start].match != NIL or self.nodes[start].min_level != 0:
                continue

            queue = [start]
            pred = [NIL] * self.n
            vis = [False] * self.n
            vis[self.find_base(start)] = True
            qi = 0
            endpoint = NIL

            while qi < len(queue) and endpoint == NIL:
                u = queue[qi]
                qi += 1

                for v in self.graph[u]:
                    bu = self.find_base(u)
                    bv = self.find_base(v)
                    if bu == bv or vis[bv]:
                        continue

                    if self.nodes[v].match == NIL and v != start:
                        pred[v] = u
                        endpoint = v
                        break

                    pred[v] = u
                    vis[bv] = True
                    mv = self.nodes[v].match
                    if mv != NIL and not vis[self.find_base(mv)]:
                        pred[mv] = v
                        vis[self.find_base(mv)] = True
                        queue.append(mv)

            if endpoint != NIL:
                path = []
                curr = endpoint
                while curr != NIL:
                    path.append(curr)
                    curr = pred[curr]
                path.reverse()

                i = 0
                while i + 1 < len(path):
                    self.nodes[path[i]].match = path[i + 1]
                    self.nodes[path[i + 1]].match = path[i]
                    i += 2
                found = True

        return found

    def maximum_matching(self):
        while True:
            self.phase_1()
            if not self.phase_2():
                break

        matching = []
        for u in range(self.n):
            if self.nodes[u].match != NIL and self.nodes[u].match > u:
                matching.append((u, self.nodes[u].match))
        matching.sort()
        return matching


def validate_matching(n, graph, matching):
    deg = [0] * n
    errors = 0

    for u, v in matching:
        pos = bisect_left(graph[u], v)
        if pos >= len(graph[u]) or graph[u][pos] != v:
            print(f"ERROR: Edge ({u}, {v}) not in graph!", file=sys.stderr)
            errors += 1
        deg[u] += 1
        deg[v] += 1

    for i in range(n):
        if deg[i] > 1:
            print(f"ERROR: Vertex {i} in {deg[i]} edges!", file=sys.stderr)
            errors += 1

    matched = sum(1 for d in deg if d > 0)

    print()
    print("=== Validation Report ===")
    print(f"Matching size: {len(matching)}")
    print(f"Matched vertices: {matched}")
    print("VALIDATION FAILED" if errors > 0 else "VALIDATION PASSED")
    print("=========================")
    print()


def load_graph(filename):
    with open(filename, 'r') as f:
        n, m = map(int, f.readline().split())
        edges = []
        for line in f:
            parts = line.split()
            if len(parts) >= 2:
                edges.append((int(parts[0]), int(parts[1])))
    return n, edges


def main():
    print("Micali-Vazirani Algorithm (Hybrid) - Python Implementation")
    print("=============================================================")
    print()

    if len(sys.argv) < 2:
        print(f"Usage: python {sys.argv[0]} <filename>")
        sys.exit(1)

    n, edges = load_graph(sys.argv[1])
    print(f"Graph: {n} vertices, {len(edges)} edges")

    t0 = time.time()
    mv = MicaliVazirani(n, edges)
    matching = mv.maximum_matching()
    t1 = time.time()

    validate_matching(n, mv.graph, matching)

    print(f"Matching size: {len(matching)}")
    print(f"Time: {int((t1 - t0) * 1000)} ms")


if __name__ == "__main__":
    main()

File: algorithms/micali-vazirani/rust/micali_vazirani.rs
/*
 * Micali-Vazirani Algorithm (Hybrid) - O(E√V) Maximum Matching
 *
 * Hybrid approach:
 * - MV-style MIN phase (level building with even/odd tracking)
 * - Gabow-style MAX phase (BFS path finding and augmentation)
 *
 * Rust implementation — fully deterministic, no hash containers.
 */

use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;

const NIL: i32 = -1;
const UNSET: i32 = i32::MAX;

struct Node {
    preds: Vec<usize>,
    match_: i32,
    min_level: i32,
    even_level: i32,
    odd_level: i32,
}

impl Node {
    fn new() -> Self {
        Node { preds: Vec::new(), match_: NIL, min_level: UNSET, even_level: UNSET, odd_level: UNSET }
    }

    fn set_min_level(&mut self, level: i32) {
        self.min_level = level;
        if level % 2 == 0 { self.even_level = level; }
        else { self.odd_level = level; }
    }

    fn reset(&mut self) {
        self.preds.clear();
        self.min_level = UNSET;
        self.even_level = UNSET;
        self.odd_level = UNSET;
    }
}

struct MicaliVazirani {
    n: usize,
    greedy_size: usize,
    graph: Vec<Vec<usize>>,
    nodes: Vec<Node>,
    base: Vec<usize>,
    levels: Vec<Vec<usize>>,
}

impl MicaliVazirani {
    fn new(n: usize, edges: &[(usize, usize)]) -> Self {
        let mut graph = vec![Vec::new(); n];
        for &(u, v) in edges {
            if u < n && v < n && u != v {
                graph[u].push(v);
                graph[v].push(u);
            }
        }
        for adj in &mut graph { adj.sort_unstable(); adj.dedup(); }

        MicaliVazirani {
            n,
            greedy_size: 0,
            graph,
            nodes: (0..n).map(|_| Node::new()).collect(),
            base: (0..n).collect(),
            levels: Vec::new(),
        }
    }

    fn find_base(&mut self, v: usize) -> usize {
        if self.base[v] != v {
            self.base[v] = self.find_base(self.base[v]);
        }
        self.base[v]
    }

    fn add_to_level(&mut self, level: usize, node: usize) {
        while self.levels.len() <= level { self.levels.push(Vec::new()); }
        self.levels[level].push(node);
    }

    fn step_to(&mut self, to: usize, from: usize, level: i32) {
        let next = level + 1;
        let tl = self.nodes[to].min_level;
        if tl >= next {
            if tl != next {
                self.add_to_level(next as usize, to);
                self.nodes[to].set_min_level(next);
            }
            self.nodes[to].preds.push(from);
        }
    }

    fn phase_1(&mut self) {
        self.levels.clear();
        for i in 0..self.n { self.base[i] = i; self.nodes[i].reset(); }

        for i in 0..self.n {
            if self.nodes[i].match_ == NIL {
                self.add_to_level(0, i);
                self.nodes[i].set_min_level(0);
            }
        }

        for i in 0..self.n {
            if i >= self.levels.len() || self.levels[i].is_empty() { continue; }
            let level_snap = self.levels[i].clone();
            for &cur in &level_snap {
                let match_ = self.nodes[cur].match_;
                if i % 2 == 0 {
                    let neighbors = self.graph[cur].clone();
                    for nb in neighbors {
                        if nb as i32 != match_ { self.step_to(nb, cur, i as i32); }
                    }
                } else {
                    if match_ != NIL { self.step_to(match_ as usize, cur, i as i32); }
                }
            }
        }
    }

    fn phase_2(&mut self) -> bool {
        let mut found = false;

        for start in 0..self.n {
            if self.nodes[start].match_ != NIL || self.nodes[start].min_level != 0 { continue; }

            let mut queue = vec![start];
            let mut pred = vec![NIL; self.n];
            let mut vis = vec![false; self.n];
            let mut qi = 0;

            vis[self.find_base(start)] = true;
            let mut endpoint: Option<usize> = None;

            while qi < queue.len() && endpoint.is_none() {
                let u = queue[qi];
                qi += 1;

                let neighbors = self.graph[u].clone();
                for v in neighbors {
                    let bu = self.find_base(u);
                    let bv = self.find_base(v);
                    if bu == bv || vis[bv] { continue; }

                    if self.nodes[v].match_ == NIL && v != start {
                        pred[v] = u as i32;
                        endpoint = Some(v);
                        break;
                    }

                    pred[v] = u as i32;
                    vis[bv] = true;
                    let mv = self.nodes[v].match_;
                    if mv != NIL && !vis[self.find_base(mv as usize)] {
                        pred[mv as usize] = v as i32;
                        vis[self.find_base(mv as usize)] = true;
                        queue.push(mv as usize);
                    }
                }
            }

            if let Some(mut curr) = endpoint {
                let mut path = Vec::new();
                loop {
                    path.push(curr);
                    if pred[curr] == NIL { break; }
                    curr = pred[curr] as usize;
                }
                path.reverse();

                let mut i = 0;
                while i + 1 < path.len() {
                    self.nodes[path[i]].match_ = path[i + 1] as i32;
                    self.nodes[path[i + 1]].match_ = path[i] as i32;
                    i += 2;
                }
                found = true;
            }
        }
        found
    }


    fn greedy_init(&mut self) -> usize {
        let mut cnt: usize = 0;
        for u in 0..self.n {
            if self.mate[u] != NIL { continue; }
            let neighbors: Vec<usize> = self.graph[u].clone();
            for &v in &neighbors {
                if self.mate[v] == NIL {
                    self.mate[u] = v as i32;
                    self.mate[v] = u as i32;
                    cnt += 1;
                    break;
                }
            }
        }
        cnt
    }

    /* Min-degree greedy: match each exposed vertex with its lowest-degree unmatched neighbor */
    fn greedy_init_md(&mut self) -> usize {
        let mut cnt: usize = 0;
        let mut deg = vec![0usize; self.n];
        for u in 0..self.n {
            for &v in &self.graph[u] {
                deg[v] += 1;
            }
        }
        let mut order: Vec<usize> = (0..self.n).collect();
        order.sort_unstable_by_key(|&u| deg[u]);
        for u in order {
            if self.mate[u] != NIL { continue; }
            let mut best: i32 = -1;
            let mut best_deg = usize::MAX;
            let neighbors: Vec<usize> = self.graph[u].clone();
            for &v in &neighbors {
                if self.mate[v] == NIL && deg[v] < best_deg {
                    best = v as i32;
                    best_deg = deg[v];
                }
            }
            if best >= 0 {
                self.mate[u] = best;
                self.mate[best as usize] = u as i32;
                cnt += 1;
            }
        }
        cnt
    }

    fn maximum_matching(&mut self, greedy_mode: i32) -> Vec<(usize, usize)> {
        self.greedy_size = match greedy_mode {
            1 => self.greedy_init(),
            2 => self.greedy_init_md(),
            _ => 0,
        };
        loop {
            self.phase_1();
            if !self.phase_2() { break; }
        }

        let mut matching = Vec::new();
        for u in 0..self.n {
            if self.nodes[u].match_ != NIL && (self.nodes[u].match_ as usize) > u {
                matching.push((u, self.nodes[u].match_ as usize));
            }
        }
        matching.sort_unstable();
        matching
    }
}

fn validate_matching(n: usize, graph: &[Vec<usize>], matching: &[(usize, usize)]) {
    let mut deg = vec![0i32; n];
    let mut errors = 0;

    for &(u, v) in matching {
        if !graph[u].binary_search(&v).is_ok() {
            eprintln!("ERROR: Edge ({}, {}) not in graph!", u, v);
            errors += 1;
        }
        deg[u] += 1;
        deg[v] += 1;
    }
    for i in 0..n {
        if deg[i] > 1 { eprintln!("ERROR: Vertex {} in {} edges!", i, deg[i]); errors += 1; }
    }
    let matched = deg.iter().filter(|&&d| d > 0).count();

    println!("\n=== Validation Report ===");
    println!("Matching size: {}", matching.len());
    println!("Matched vertices: {}", matched);
    println!("{}", if errors > 0 { "VALIDATION FAILED" } else { "VALIDATION PASSED" });
    println!("=========================\n");
}

fn load_graph(filename: &str) -> Result<(usize, Vec<(usize, usize)>), Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    let first = lines.next().ok_or("Empty file")??;
    let parts: Vec<&str> = first.split_whitespace().collect();
    let n: usize = parts[0].parse()?;
    let m: usize = parts[1].parse()?;

    let mut edges = Vec::with_capacity(m);
    for line in lines {
        let line = line?;
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 {
            let u: usize = parts[0].parse()?;
            let v: usize = parts[1].parse()?;
            edges.push((u, v));
        }
    }
    Ok((n, edges))
}

fn main() {
    println!("Micali-Vazirani Algorithm (Hybrid) - Rust Implementation");
    println!("==========================================================\n");

    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <filename> [--greedy|--greedy-md]", args[0]);
        std::process::exit(1);
    }

    let greedy_mode: i32 = if args.iter().any(|a| a == "--greedy-md") { 2 } else if args.iter().any(|a| a == "--greedy") { 1 } else { 0 };
    match load_graph(&args[1]) {
        Ok((n, edges)) => {
            println!("Graph: {} vertices, {} edges", n, edges.len());

            let start = Instant::now();
            let mut mv = MicaliVazirani::new(n, &edges);
            let matching = mv.maximum_matching(greedy_mode);
            let duration = start.elapsed();

            validate_matching(n, &mv.graph, &matching);

            println!("Matching size: {}", matching.len());
            if greedy_mode > 0 {
                let gs = mv.greedy_size;
                let fs = matching.len();
                println!("Greedy init size: {}", gs);
                if fs > 0 { println!("Greedy/Final: {:.2}%", 100.0 * gs as f64 / fs as f64); }
                else { println!("Greedy/Final: NA"); }
            }
            println!("Time: {} ms", duration.as_millis());
        }
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    }
}
